weave/trace_server/execution_runner/cross_process_trace_server.py

```diff
@@ -0,0 +1,540 @@
+        """Replace feedback."""
+        return self._send_request("feedback_replace", req)
+
+    # === Action API ===
+    def actions_execute_batch(
+        self, req: ActionsExecuteBatchReq
+    ) -> ActionsExecuteBatchRes:
+        """Execute actions in batch."""
+        return self._send_request("actions_execute_batch", req)
+
+    # === Execute LLM API ===
+    def completions_create(self, req: CompletionsCreateReq) -> CompletionsCreateRes:
+        """Create LLM completions."""
+        return self._send_request("completions_create", req)
+
+    def completions_create_stream(
+        self, req: CompletionsCreateReq
+    ) -> Iterator[dict[str, Any]]:
+        """
+        Create LLM completions as a stream.
+
+        Note: Streaming is complex across threads. Currently falls back
+        to non-streaming version.
+        """
+        # TODO: Implement proper streaming support
+        result = self.completions_create(req)
+        yield result.response
+
+    # === Project statistics API ===
+    def project_stats(self, req: ProjectStatsReq) -> ProjectStatsRes:
+        """Get project statistics."""
+        return self._send_request("project_stats", req)
+
+    # === Thread API ===
+    def threads_query_stream(self, req: ThreadsQueryReq) -> Iterator[ThreadSchema]:
+        """
+        Query threads as a stream.
+
+        Note: This would need special handling for streaming.
+        Currently not implemented.
+        """
+        # TODO: Implement streaming support
+        raise NotImplementedError(
+            "Streaming not yet implemented for cross-process server"
+        )
+
+    # === Evaluation Execution API ===
+    async def run_model(self, req: RunModelReq) -> RunModelRes:
+        """
+        Run a model asynchronously.
+
+        Note: Async methods require special handling. Currently runs synchronously.
+        """
+        # TODO: Implement proper async support
+        return self._send_request("run_model", req)
+
+
+def generate_child_process_trace_server_args(
+    trace_server: TraceServerInterface,
+) -> CrossProcessTraceServerArgs:
+    """
+    Start a worker thread with the trace server and return serializable args.
+
+    This function runs in the PARENT process and:
+    1. Creates multiprocessing queues for communication
+    2. Captures current context variables
+    3. Starts a worker thread that runs the given trace server with context
+    4. Returns ONLY the queues (which are serializable)
+
+    The trace server stays in the parent process, only the queues are passed to child.
+
+    Args:
+        trace_server: The trace server instance to run in the worker thread (stays in parent)
+
+    Returns:
+        A dictionary containing the queues to pass to the child process
+    """
+    # Get the multiprocessing context
+    ctx = multiprocessing.get_context()
+
+    # Create multiprocessing queues (these CAN be serialized)
```

Comment:
and are safe to write to between processes with impunity. no need to take a lock

Reply:
correct - the queue mechanism is designed to handle multiple process writing to the queue